fe310: Add custom PLIC driver as peripheral

bergzand opened the pull request:
### Contribution description

This PR adds a custom driver for the RISC-V standardized platform-local interrupt controller. This removes vendor-supplied code in favor of our own driver. The PLIC here is fully optional except for fe310-specific peripheral drivers. This allows for swapping the PLIC driver for a different interrupt controller on other RISC-V platforms.

### Testing procedure

Test if stdin still works (uart interrupt), the RTT test and a GPIO test must also still work.

### Issues/PRs references

Nothing yet, but this makes the RISC-V PLIC a bit more modular in preparation for the gd32v, which uses an ECLIC for the interrupts.

aabadie commented on cpu/fe310/periph/plic.c:
```diff
@@ -0,0 +1,121 @@
+
+    volatile uint32_t *irq_reg = &PLIC_REG(PLIC_ENABLE_OFFSET +
+                                           (hart_id <<
+    PLIC_ENABLE_SHIFT_PER_TARGET) +
```
indent is off here. Same in `plic_disable_interrupt`

bergzand replied:
Blame uncrustify :disappointed: 

I reworked it into a separate register getter for the `irq_reg`, de-duplicating the register address magic in the process.

aabadie commented on cpu/fe310/include/cpu.h:
```diff
@@ -81,6 +71,7 @@ static inline void cpu_print_last_instruction(void)
  */
 void nanostubs_init(void);
 
+
```
unrelated change

bergzand replied:
removed

aabadie reviewed (commented):
Nice one! Code looks ok expect some minor style thing. I'll test that later.

bergzand commented:
This is currently broken on the hifive1 boards. Somehow it does work on the sifive_e Qemu machine. Could be that some of the register calculations are not correct, will check tomorrow.

bergzand commented on cpu/fe310/periph/plic.c:
```diff
@@ -0,0 +1,121 @@
+
+    return &PLIC_REG(PLIC_ENABLE_OFFSET +
+                     (hart_id << PLIC_ENABLE_SHIFT_PER_TARGET) +
+                     (irq >> 5));
```
I managed to sneak in an unaligned access here, This should progress to the next word for every 32 interrupts, but this increases the address by 1 every 32 `irq`'s and not by 4.

bergzand commented:
> This is currently broken on the hifive1 boards. Somehow it does work on the sifive_e Qemu machine. Could be that some of the register calculations are not correct, will check tomorrow.

As mentioned above. The register calculation for the IRQ enable/disable register was incorrect, incrementing the address by 1 for every 32 IRQ lines and not by 4. This caused an unaligned access to the register, triggering a fault. Fixed it and the code should be working properly again.

I've also replaced the bit operations in the irq enable/disable with GCC atomics. These compile to the atomic bit operations supported on the rv32im**a**c architecture. Setting the mode to `__ATOMIC_RELAXED` assembles the atomic operations in such a way that the RISC-V hart doesn't have to add the multiprocessor safety features with the instructions.

aabadie reviewed (approved):
Tested on hifive1b `tests/preiph_rtt`, `tests/stdin` and tested GPIO interrupt using a custom hello-world (with a pin configured as input). All is working OK.

Code changes are good as well.

ACK

aabadie commented:
please squash!

bergzand commented:
Squashed!
